### ISSUE TITLE
Use the new Repositories API in the UI

### ISSUE THREAD

**Issue by antgamdia** [vmware-tanzu/kubeapps#4764]
As the new API for managing Packages Repositories is becoming more stable, the current UI needs to be properly adapted in consequence.

Currently, we can test it with the Flux plugin, but there is ongoing work on both the Helm (#4662) and Carvel (#4681) plugins.

IMO, the new design should be carried out in multiple stages:

- Phase 0: https://github.com/vmware-tanzu/kubeapps/pull/4817
    - [x]  Start implementing towards that direction (autogenerated apprepo grpc clients, start using new API calls, report issues back, improve the API if needed, etc)
    - [x] Ensure feature parity between the new API and what Kubeapps already supports.
    - [x] Ensure the implementation is consistent across plugins.
    - PRs:
      - https://github.com/vmware-tanzu/kubeapps/pull/4930
      - https://github.com/vmware-tanzu/kubeapps/pull/4939
      - https://github.com/vmware-tanzu/kubeapps/pull/4954
      - https://github.com/vmware-tanzu/kubeapps/pull/4955
- Phase 1:  
   - Blocking issues:
   -  [x] Allow private global repos in the Helm plugin
      - Tracked at: https://github.com/vmware-tanzu/kubeapps/issues/5129
   -  [x] Re-enable e2e tests: https://github.com/vmware-tanzu/kubeapps/issues/4876
      - Done at https://github.com/vmware-tanzu/kubeapps/pull/5115
      - Blocked by https://github.com/vmware-tanzu/kubeapps/issues/5129
    - [x] Add a way to specify img pull secrets in the helm plugin for kubeapps-managed secrets. Currently, the UI supports both approaches, but the API only allows setting the secret manually. This has led to disabling the e2e test nº 03/10.
      - Tracked at: https://github.com/vmware-tanzu/kubeapps/issues/5015
      - Done at:
        - https://github.com/vmware-tanzu/kubeapps/pull/5081
        - https://github.com/vmware-tanzu/kubeapps/pull/5109
    - [x] Add the `accessLevel` (= type of auth) to the  PackageRepositorySummary API + use it in the UI
      - Trivial, done at: https://github.com/vmware-tanzu/kubeapps/pull/5003
    - [x]  The helm plugin doesn't have a status.ready
      - Trivial, done at: https://github.com/vmware-tanzu/kubeapps/pull/5006
    - [x] The auth type "Bearer" implementation is inconsistent across the plugins
      - Tracked at: https://github.com/vmware-tanzu/kubeapps/issues/5028
      - Done at: https://github.com/vmware-tanzu/kubeapps/pull/5097
   - Improvements:
    - [x] Clean up the unused `IPackageRepositoryState` and its actions.
      - Tracked at: https://github.com/vmware-tanzu/kubeapps/issues/5061
      - Done at: https://github.com/vmware-tanzu/kubeapps/pull/5062
    - [x] Rename `RepositoryCustomDetails` to `HelmRepositoryCustomDetails` or `RepositoryHelmCustomDetails`. Idem with Carvel.
      - Tracked at: https://github.com/vmware-tanzu/kubeapps/issues/5016
      - Done at: https://github.com/vmware-tanzu/kubeapps/pull/5046
    - [x] Delete `url.apprepositories.get` once we don't need it anymore. It also means we will be able to remove the endpoints from kubeops!
      - Tracked at: https://github.com/vmware-tanzu/kubeapps/issues/5014

**Comment by antgamdia** [comment]
> Phase 0:

I've been working these days on this thing, mostly ensuring that what we currently have is actually supported by the new API.

Just writing down some thoughts during the (still ongoing) implementation:

#### General
- No validation endpoint is present. Ideas? 
    - Add it as a core /validate endpoint
    - Add the `dryRun`: true opt-in flag to the addRepo function?
    - Add a `/validate` as a per-plugin operation (not core)?
- Access level (private/public) not in the summaries response
    - Currently, we state whether or not a repo has some sort of auth in the repo list. I don't see any harm in adding that field to the summaries response, I think? 
- Force resync operation is not available? 
    - Maybe it has been already discussed, but just noticed there is no way to force an update/re-fetch/re-sync/whatever of a given repo.
- Packages should have a field for with PackageRepositoryReference
    - Not a short-term thing, but packages API ought to be updated to add a `PackageRepositoryReference` field, no?

#### Flux

- Add "description" as a flux CR metadata?
    - Already a `todo` in the code, but I guess it can be safely implemented as an annotation in the CR?

**Comment by antgamdia** [comment]
And also... a PoC of Flux repos being managed from our current UI :)

![fluxrepo](https://user-images.githubusercontent.com/11535726/170665037-7f5aa977-1b4e-4967-8efd-1b4e8d4c3587.gif)

**Comment by ppbaena** [comment]
Wow! It looks nice 🎉

**Comment by absoludity** [comment]
Nice - looking great Antonio!

> 
>     * No validation endpoint is present. Ideas?
>       
>       * Add it as a core /validate endpoint
>       * Add the `dryRun`: true opt-in flag to the addRepo function?
>       * Add a `/validate` as a per-plugin operation (not core)?

We chatted about this last week: what is the validate function currently doing? If it is just ensuring that an http endpoint can be reached with certain credentials, we could have it as a generic function in the `resources` plugin or similar, but I don't think it is - we updated the helm plugin validation to handle OCI repos as well (which did some OCI-related validation).

For 1-1 with the current api, I'd assume we'd want to add a validation endpoint per plugin, but not sure - what do you think?

> 
>     * Access level (private/public) not in the summaries response
>       
>       * Currently, we state whether or not a repo has some sort of auth in the repo list. I don't see any harm in adding that field to the summaries response, I think?

I don't see a problem with that. Note: don't think we should conflate the presence of auth with private/public .
 
>     * Force resync operation is not available?       
>       * Maybe it has been already discussed, but just noticed there is no way to force an update/re-fetch/re-sync/whatever of a given repo.

That's because it's not something that makes sense generally (eg. what would this mean for a Carvel repository? I'm not sure we can tell the carvel controller to resync, it may not even makes sense because the carvel repo version is meant to be frozen?). If we add this for the helm repository, it'll be an extra RPC added to the helm implementation (ie. not in the core) and any other plugin that supports it, but not sure we need to do that initially. WDYT?

> 
>     * Packages should have a field for with PackageRepositoryReference
>       
>       * Not a short-term thing, but packages API ought to be updated to add a `PackageRepositoryReference` field, no?

Confused as to which context you think is missing this? For eg, the `PackageRepositorySummary` message includes a `PackageRepositoryReference`? Or maybe I've missed what you mean here.

**Comment by antgamdia** [comment]
> We chatted about this last week: what is the validate function currently doing? If it is just ensuring that an http endpoint can be reached with certain credentials, we could have it as a generic function in the resources plugin or similar, but I don't think it is - we updated the helm plugin validation to handle OCI repos as well (which did some OCI-related validation).
>
> For 1-1 with the current api, I'd assume we'd want to add a validation endpoint per plugin, but not sure - what do you think?

As of today, I've left the existing old endpoint + only invoking it if the plugin is `helm`. During the next days, I will replace it with what Rafa implemented in the Helm plugin natively.
Later on, we can just extend the validation with a per-plugin validation endpoint, as you mentioned.

>> Access level (private/public) not in the summaries response
>
>I don't see a problem with that. Note: don't think we should conflate the presence of auth with private/public .

Great, I'll do in a folloup PR then. Adding it to this task's items.

> That's because it's not something that makes sense generally (eg. what would this mean for a Carvel repository? I'm not sure we can tell the carvel controller to resync, it may not even makes sense because the carvel repo version is meant to be frozen?). If we add this for the helm repository, it'll be an extra RPC added to the helm implementation (ie. not in the core) and any other plugin that supports it, but not sure we need to do that initially. WDYT?

Mmm, fair point. I agree, this is not a required feature right now (in fact, I've disabled the resync buttons in the initial implementation I did. However, I still see this is something useful somehow. Let's suppose I have created a repo with a sync interval of 1 day. However, someone says a new package is available and I want to try it out asap. Inituitively, I'd look for a "sync" button to kind-of-override the 1-day interval and just trigger a re-fetch in this very moment that I need it.
Anyway, let's postpone any action as there are more urgent stuff to handle (handling secrets, mainly).

> Confused as to which context you think is missing this? For eg, the PackageRepositorySummary message includes a PackageRepositoryReference? Or maybe I've missed what you mean here.

No, I meant that, currently, our packages API (like `GetAvailablePackageSummary`) is encoding the repo as part of the identifier (as in `bitnami/wordpress` --> repo=`bitnami`). What I was suggesting is adding a new field with the `PackageRepositoryReference`, but not for now.

**Comment by antgamdia** [comment]
I have some questions before going further on #4817:

Currently, the UI we have (and, we still want to maintain for the sake of this very first PR/version/phase) is like:

- Repo authorization: none, http-header auth, docker registry credentials (= user will specify the secret name)
- Image pull secrets: docker registry credentials (= user will specify the secret name)

In order to create the secret, it is the UI that handles the secret creation (see the "submit" button below)

![image](https://user-images.githubusercontent.com/11535726/171654099-8a355727-cc62-4895-8fa6-45ae679f9b18.png)


I'm now trying to figure out the mapping between our current UX and the API fields. I've sketched up a mapping (sorry for the handwriting haha):

![image](https://user-images.githubusercontent.com/11535726/171655396-0be1dca8-8857-464b-a641-b081e12796ab.png)

![image](https://user-images.githubusercontent.com/11535726/171655542-962cab69-e0c7-4922-8f35-a10107b64de8.png)


Let me explain it a little bit:

    - `auth.header`: bearer token or any http-header-based authentication.
    - `auth.passCredentials`: whether or not passing the credentials to 3rd party sites.
    - `auth.usernamePassword`: QUESTION A
    - `auth.secretRef`: QUESTION B
    - `auth.tlsCertKey`: QUESTION C
    - `tlsConfig: certAuthority`: our current "custo CA certificate"
    - `tlsConfig: insecureSkipVerify`: our current skip tls verification
    - `tlsConfig.secretRef`: QUESTION D

1. Question A: if user/password refers to an HTTP Basic Authentication (meaning header like `"Basic "+base64(user+":"+password)`), why would we need the `usernamePassword` object? Perhaps to make it easier for clients? Can we just ignore it and use the `header` field?
2. Question B: if it is the UI that is creating the secret, should we pass the secret name here? I mean, assuming there is a secret `my-secret` in the "associate docker registry credentials" dropdown AND user selects `use docker registry credentials`. Am I right? 
    - Set `auth.secretRef.key` to `my-secret`
    - Set `auth.secretRef.name` to `???????`: 
3. Question C: `auth.tlsCertKey` always unset, we don't support it right now, do we?
4. Question D: `tlsConfig.secretRef` if user selects a secret in the `use docker registry credentials` dropdown, populate it as follows. Am I right?
    - Set `tlsConfig.secretRef.key` to `my-secret`
    - Set `tlsConfig.secretRef.name` to `???????`: 

In that scenario, I guess we are just supporting the `user-managed secrets` ([see discussion on Slack](https://kubernetes.slack.com/archives/C9D3TSUG4/p1654159619865549)), no?

**Comment by antgamdia** [comment]
> A) Only advantage I see to having a separate data structure for u/p is that we will be able to show REDACTED only on the password and not the name. If we have u/p in the header, the whole header value will show REDACTED I guess. Leaving the UI user without knowing which user is being used for authentication. 

Great, added to this issue to add it in a follow-up PR.

> B) The UI will not create the secret with Package repository API, it will be the backend. Here is where the UI could have some toggle for selecting whether if Kubeapps will manage secrets or not. 

Ah, ok. Yep, I meant the Secret is being created by the resources api, but triggered by the UI. As opposed as the repos API creating it.
For my ongoing PR I plan to leave it as is, so that we can re-think it and add the toggle for user/kubeapps-managed secrets.
 
> C)  `auth.tlsCertKey` at least for Helm we do not support it AFAIK

Great, thx.

> D) If user selects Docker registry credentials secret, that would go under `auth.secretRef`, no? The `tlsConfig.secretRef` is for the secret of the Custom CA? 

Yep, you're totally right. Thanks for pointing it out!


> IMHO in order to make the UI more useful and given that we are restricted to set the "kubeapps manages secrets" flag per installation by now, I would offer the "kubeapps manages secrets" mode by now. This is, no secrets picking from UI. In this way, the whole thing can be done from UI. WDYT?

+1 to this idea, but not right now. My ongoing PR has several changes and I'd prefer not to add any more disruptive changes here but in a follow-up PR. I mean, getting rid of the secret selector will have an inpact on the test suites :S

I'm going to polish up my PR and extract any pending TODO from it soon.

**Comment by castelblanque** [comment]
Thanks for the comments. Just a remark:

> Yep, I meant the Secret is being created by the resources api, but triggered by the UI. As opposed as the repos API creating it.

With the new API secrets are created inside the add/update repository operation (when "kubeapps manages secrets" mode). Not created by a previous API invocation to the resources API.

**Comment by absoludity** [comment]
> Currently, the UI we have (and, we still want to maintain for the sake of this very first PR/version/phase) is like:

I don't think it will be possible to maintain the existing UI when switching to the new API. Because we already wanted to simplify the UX for app repositories, the API has been designed not to necessarily support an existing UX, but rather, to support.a simplified interaction.

**Comment by antgamdia** [comment]
Ok, so perhaps a good approach is to leave the current PR as-is (current UI, falling tests) and, on top of it, start stacking the rest of the follow-up PRs removing the current logic for handling secrets manually and start fixing the test cases in this new "simplified" UI.
Thanks for your comments!

**Comment by antgamdia** [comment]
FTR (from https://github.com/vmware-tanzu/kubeapps/issues/4361#issue-1155032753):

> This issue is a by-product of #3496 The core apis server now has support for RepositoriesService, which currently only has one API [AddPackageRepository](https://docs.google.com/document/d/1z03msZRGsRvcRaom-yrvEwIWcEfNy6fSA5Zg28gjYvA/edit#heading=h.u3lj7whdof7t)(). Only flux plugin currently implements the API. There will soon be more APIs defined for Reading, Updaing and Deleting of repositories So:
> 
> 1. The dashboard needs to be refactored to start to use the new Repositories Service. We should start with 'Add...' and move on to other operations as they become available
> 2. the Helm plugin and carvel plugins need to implement the AddPackageRepository() API (including the custom field handling the plugin specific stuff)

**Comment by antgamdia** [comment]
I have extracted some of the longer-term items into a separate issue: https://github.com/vmware-tanzu/kubeapps/issues/5121
